lib/Sema/CSFix.cpp

```diff
@@ -2151,3 +2151,127 @@ IgnoreDefaultExprTypeMismatch::create(ConstraintSystem &cs, Type argType,
+        &openedExistentials,
+    ConstraintLocatorBuilder locator) {
+  using ExistentialList =
+      SmallVectorImpl<std::pair<TypeVariableType *, OpenedArchetypeType *>> &;
+
+  struct CoercionChecker : public TypeWalker {
+    bool RequiresCoercion = false;
+
+    ConstraintSystem &cs;
+    ExistentialList OpenedExistentials;
+
+    CoercionChecker(ConstraintSystem &cs, ExistentialList openedExistentials)
+        : cs(cs), OpenedExistentials(openedExistentials) {}
+
+    Action walkToTypePre(Type componentTy) override {
+      // In cases where result references a member type, we need to check
+      // whether such type would is resolved to concrete or not.
+      if (auto *member = componentTy->getAs<DependentMemberType>()) {
+        Type memberBaseTy = member->getBase();
+
+        // Find the base of this member chain.
+        while (true) {
+          if (auto *memberTy = memberBaseTy->getAs<DependentMemberType>()) {
+            memberBaseTy = memberTy->getBase();
+          } else {
+            break;
+          }
+        }
+
+        auto typeVar = memberBaseTy->getAs<TypeVariableType>();
+        if (!typeVar)
+          return Action::SkipChildren;
+
+        // If the base is an opened existential type, let's see whether
+        // erase would produce an existential in this case and if so,
+        // we need to check whether any requirements are going to be lost
+        // in process.
+
+        auto opened =
+            llvm::find_if(OpenedExistentials, [&typeVar](const auto &entry) {
+              return typeVar == entry.first;
+            });
+
+        if (opened == OpenedExistentials.end())
+          return Action::SkipChildren;
+
+        auto erasedMemberTy = typeEraseOpenedExistentialReference(
+            Type(member), opened->second->getExistentialType(), opened->first,
+            TypePosition::Covariant);
+
+        // If result is an existential type and the base has `where` clauses
+        // associated with its associated types, the call needs a coercion.
+        if (erasedMemberTy->isExistentialType() &&
+            hasConstrainedAssociatedTypes(opened->second)) {
+          RequiresCoercion = true;
+          return Action::Stop;
+        }
+
+        return Action::SkipChildren;
+      }
+
+      // The case where there is a direct access to opened existential type
+      // e.g. `$T` or `[$T]`.
+      if (auto *typeVar = componentTy->getAs<TypeVariableType>()) {
+        auto opened =
+            llvm::find_if(OpenedExistentials, [&typeVar](const auto &entry) {
+              return typeVar == entry.first;
+            });
+
+        if (opened != OpenedExistentials.end()) {
+          RequiresCoercion |= hasConstrainedAssociatedTypes(opened->second);
+          return RequiresCoercion ? Action::Stop : Action::SkipChildren;
+        }
+      }
+
+      return Action::Continue;
+    }
+
+  private:
+    bool hasConstrainedAssociatedTypes(ArchetypeType *archetypeTy) {
```

Comment:
I realized that this condition is incorrect, I need to check whether any member in the chain of associated types described by dependent member ends up having a where clause.

Reply:
Fixed!